core/src/connection/pool.rs

```diff
@@ -0,0 +1,907 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{
+    Executor,
+    ConnectedPoint,
+    PeerId,
+    connection::{
+        self,
+        ConnectionId,
+        ConnectionLimit,
+        IncomingInfo,
+        OutgoingInfo,
+        Connection,
+        Substream,
+        Connected,
+        ConnectionError,
+        ConnectionHandler,
+        IntoConnectionHandler,
+        ConnectionInfo,
+        manager::{self, Manager},
+    },
+    muxing::StreamMuxer,
+};
+use either::Either;
+use fnv::FnvHashMap;
+use futures::prelude::*;
+use smallvec::SmallVec;
+use std::{error, fmt, hash::Hash, task::Context, task::Poll};
+
+/// A connection `Pool` manages a set of connections for each peer.
+pub struct Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo = PeerId, TPeerId = PeerId> {
+    local_id: TPeerId,
+
+    /// The configuration of the pool.
+    limits: PoolLimits,
+
+    /// The connection manager that handles the connection I/O for both
+    /// established and pending connections.
+    ///
+    /// For every established connection there is a corresponding entry in `established`.
+    manager: Manager<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo>, // tasks
+
+    /// The managed connections of each peer that are currently considered
+    /// established, as witnessed by the associated `ConnectedPoint`.
+    established: FnvHashMap<TPeerId, FnvHashMap<ConnectionId, ConnectedPoint>>,
+
+    /// The pending connections that are currently being negotiated.
+    pending: FnvHashMap<ConnectionId, (ConnectedPoint, Option<TPeerId>)>,
+}
+
+impl<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> fmt::Debug
+for Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        // TODO: More useful debug impl?
+        f.debug_struct("Pool")
+            .field("limits", &self.limits)
+            .finish()
+    }
+}
+
+impl<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> Unpin
+for Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> {}
+
+/// Event that can happen on the `Pool`.
+pub enum PoolEvent<'a, TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> {
+    /// A new connection has been established.
+    ///
+    /// In order for the connection to remain in the pool, it must be accepted.
```

Comment:
I believe this comment is no longer true?

Reply:
Correct, thanks!